### DISCUSSION TITLE
doc: @types/node JSDoc integration - guidance and call to collaboration

### DISCUSSION THREAD

**Discussion by SimonSchick** [nodejs/help#4866] [Ideas]
## Rationale

NodeJS docs are currently not well integrated into any IDEs which generally slows down developers as they have to consult the official documentation website (which may go down) or an offline copy, a mental context switch is required when wanting to look up even small details such as a functions description/behaviour.

@types/node has made NodeJS application development significantly faster, safer and easier (minus initial setup 😃 ) due to TypeScript's inherit properties and has already resolved a good amount of the aforementioned context-switches, I however believe there is still a lot of potential to improve this.

Documentation integration and upkeep in the type definitions has been a major chore and was often done very inconsistently (I'm very guilty of this).

In order to improve the development experience and reduce the reliance of developers on mutliple sources NodeJS documentation can and should be more tightly coupled/integrated with the current `@types/node` distribution.

# Work

Over the last couple of weeks I've invested a fair amount of time writing some scripts that incorporate a very large amount of existing NodeJS documentation into the current TypeScript definition.

The current progress of this effort can be inspected here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54417

# Asks

- Guidance on how/when credit/**license** should be give the copying of others people work is quite substantial.
- Invite people to **collaborate** and get input on the current integration, the current implementation is rather simplistic and rather crude.
- Discuss the **structural issues** of the current JSON doc output as I've [had to perform a lot of quirky rewrites and fixes](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54417/files#diff-d754009e48fd847158b1b90709626491fc0cc474fd3601e4e9c0d9e3d2806b9dR104) of the JSON doc in order to be able to work with it.

**Comment by Trott** [comment]
> * Discuss the **structural issues** of the current JSON doc output as I've [had to perform a lot of quirky rewrites and fixes](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/54417/files#diff-d754009e48fd847158b1b90709626491fc0cc474fd3601e4e9c0d9e3d2806b9dR104) of the JSON doc in order to be able to work with it.

Fixes for the JSON doc output are welcome. You can PR them in here. Or open specific issues. The code that generates the JSOn docs is in `tools/doc/json.mjs`.

**Reply by SimonSchick** [reply]
I vaguely remember some efforts of changing the current format to something that is used by another package, maybe electron(?) but cannot find the associated thread/issue.
Do you perhaps have a link to that?

I want to avoid working on something only for it to be tossed soon after, once I got the initial aforementioned PR in DT out I will look into fixing some of the structural un-soundness of the current docs (iirc they boil down to markdown formatting issues.

**Reply by Trott** [reply]
> I vaguely remember some efforts of changing the current format to something that is used by another package, maybe electron(?) but cannot find the associated thread/issue.
> Do you perhaps have a link to that?

https://github.com/nodejs/node/issues/32206

**Comment by Trott** [comment]
> * Guidance on how/when credit/**license** should be give the copying of others people work is quite substantial.

I'm not sure if you're talking about Node.js docs or other work, but if Node.js core stuff, it's all MIT licensed so the licensing and even crediting should be pretty straightforward.

**Reply by SimonSchick** [reply]
I'm a little out of the loop/practise but would it be sufficient to copy/insert [this section](https://github.com/nodejs/node/blob/3ac223ed25db4e4b89799f3e60ad8dd6ce454884/LICENSE#L4-L22) into the index.d.ts file that is shipped?

**Reply by Trott** [reply]
I am not a lawyer so this is not legal advice blah blah blah meow meow meow, but yes, I am under the impression that including that notice would be sufficient.

**Comment by Trott** [comment]
> * Invite people to **collaborate** and get input on the current integration, the current implementation is rather simplistic and rather crude.

You've probably noticed this, but just in case: There has been an effort lately to add JSDoc typings for the Node.js internal modules, so that might be useful. I believe @VoltrexMaster has been doing much (most? all?) of the work there.

**Reply by SimonSchick** [reply]
I saw a couple of PRs but I'm uncertain if they had any impact on doc generation so far, I will look into this!
